src/utils/crypto.ts

```diff
@@ -0,0 +1,148 @@
+  const key = await subtleCrypto.deriveKey(
+    {
+      name: 'PBKDF2',
+      salt,
+      iterations: 100000,
+      hash: 'SHA-256',
+    },
+    keyMaterial,
+    { name: PASSWORD_ENCRYPTION_ALGO, length: 256 },
+    IS_EXTRACTABLE,
+    ['encrypt', 'decrypt'],
+  );
+
+  return key;
+}
+
+export async function restoreEncryptionData(
+  password:string,
+  encryptedMnemonic: string,
+): Promise<IEncryptionData> {
+  const data = new Uint8Array(decodeBase64(encryptedMnemonic));
+  // Extract salt and iv from the data
+  const salt = data.slice(0, 16);
+  const iv = data.slice(16, 32);
+  const key = await generateCryptoKey(password, salt);
+  return { key, salt, iv };
+}
+
+export async function initializeEncryptionData(password: string): Promise<IEncryptionData> {
+  const salt = globalThis.crypto.getRandomValues(new Uint8Array(16));
+  const iv = globalThis.crypto.getRandomValues(new Uint8Array(16));
+  const key = await generateCryptoKey(password, salt);
+  return { key, salt, iv };
+}
+
+export async function encrypt(encryptionData: IEncryptionData, plaintext: string): Promise<string> {
+  const { key, salt, iv } = toRaw(encryptionData);
+  const subtleCrypto = globalThis.crypto.subtle;
+
+  // Encrypt the plaintext
+  const enc = new TextEncoder();
+  const encryptedBuffer = await subtleCrypto.encrypt(
+    {
+      name: PASSWORD_ENCRYPTION_ALGO,
+      iv,
+    },
+    key,
+    enc.encode(plaintext),
+  );
+
+  // Concatenate salt, iv, and cipher-text
+  const encryptedArray = new Uint8Array([...salt, ...iv, ...new Uint8Array(encryptedBuffer)]);
+
+  // Convert to base64
+  return encodeBase64(encryptedArray);
+}
+
+export async function decrypt(
+  encryptionData: IEncryptionData,
+  encryptedMessage: string,
+): Promise<string> {
+  const { key, iv } = toRaw(encryptionData);
+  const subtleCrypto = globalThis.crypto.subtle;
+
+  // Decode from base64
+  const encryptedArray = new Uint8Array(decodeBase64(encryptedMessage));
+  const ciphertext = encryptedArray.slice(32); // Skip salt (16 bytes) + IV (16 bytes)
+  // Decrypt the ciphertext
+  const decryptedBuffer = await subtleCrypto.decrypt(
+    {
+      name: PASSWORD_ENCRYPTION_ALGO,
+      iv,
+    },
+    key,
+    ciphertext,
+  );
+
+  // Unwrap the plaintext
+  const dec = new TextDecoder();
+  const decryptedText = dec.decode(decryptedBuffer);
+  return decryptedText;
+}
+
+export async function authenticateWithPassword(password: string): Promise<IEncryptionData> {
+  const encryptedMnemonic = await WalletStorage.get<string>(STORAGE_KEYS.mnemonic);
+  if (encryptedMnemonic) {
+    try {
+      const encryptionData = await restoreEncryptionData(password, encryptedMnemonic);
+      const decryptedMnemonic = await decrypt(encryptionData, encryptedMnemonic);
+      if (decryptedMnemonic) {
+        return Promise.resolve(encryptionData);
+      }
+      return Promise.reject(new Error('Incorrect password.'));
+    } catch (error) {
+      return Promise.reject(error);
+    }
```

Comment:
```suggestion
    const encryptionData = await restoreEncryptionData(password, encryptedMnemonic);
    const decryptedMnemonic = await decrypt(encryptionData, encryptedMnemonic);
    if (decryptedMnemonic) {
      return encryptionData;
    }
    throw new Error('Incorrect password.');
```
it makes no difference because `authenticateWithPassword` is async